python/lsst/daf/butler/core/location.py

```diff
@@ -285,37 +314,59 @@ def _fixupFileUri(parsed, root=None, forceAbsolute=False):
 
                 # normpath strips trailing "/" which makes it hard to keep
                 # track of directory vs file when calling replaceFile
-                # put it back.
+                # find the appropriate separator
                 if "scheme" in replacements:
                     sep = posixpath.sep
                 else:
                     sep = os.sep
 
-                if expandedPath.endswith(os.sep) and not replacements["path"].endswith(sep):
+                # add the trailing separator only if explicitly required or
+                # if it was stripped by normpath. Acknowledge that trailing
+                # separator exists.
+                if forceDirectory or (expandedPath.endswith(os.sep) and not
+                                      replacements["path"].endswith(sep)):
                     replacements["path"] += sep
+                    dirLike = True
 
             elif parsed.scheme == "file":
                 # file URI implies POSIX path separators so split as POSIX,
                 # then join as os, and convert to abspath. Do not handle
                 # home directories since "file" scheme is explicitly documented
                 # to not do tilde expansion.
+                sep = posixpath.sep
                 if posixpath.isabs(parsed.path):
-                    # No change needed
-                    return copy.copy(parsed)
+                    if forceDirectory:
+                        parsed = parsed._replace(path=parsed.path+sep)
+                        dirLike = True
+                    return copy.copy(parsed), dirLike
 
                 replacements["path"] = posixpath.normpath(posixpath.join(os2posix(root), parsed.path))
 
                 # normpath strips trailing "/" so put it back if necessary
-                if parsed.path.endswith(posixpath.sep) and not replacements["path"].endswith(posixpath.sep):
-                    replacements["path"] += posixpath.sep
-
+                # Acknowledge that trailing separator exists.
+                if forceDirectory or (parsed.path.endswith(sep) and not replacements["path"].endswith(sep)):
+                    replacements["path"] += sep
+                    dirLike = True
             else:
                 raise RuntimeError("Unexpectedly got confused by URI scheme")
 
             # ParseResult is a NamedTuple so _replace is standard API
             parsed = parsed._replace(**replacements)
 
-        return parsed
+        elif parsed.scheme == "s3":
```

Comment:
I believe the conditionals cover all 3 scenarios - no scheme, explicit file scheme and s3 scheme. I was afraid of assuming a posix separator as a default for any future schemes.

Reply:
I believe a `/` path separator is part of the URI standard so you can assume that -- scheme-less is special and has to be dealt with differently but everything else should be the same shouldn't it?

Reply:
Also note that this method is called fixUpFileURI so that's not a helpful name any more. It looks like the os2posix won't trigger for S3 anyhow so you can probably use the code above without a special not-file URI else.

Reply:
I removed else, renamed to fixupPathURI. Does that work for you?